### PR TITLE
Import GISAID locations, use them for Samples

### PR DESCRIPTION
**Pull request by lvreynoso**
### Summary:
- **What:** Import GISAID locations into new `locations` table; give samples a `location_id` that gives it a relationship to a `Location`; allow users to search for GISAID locations in frontend and use them when creating samples.
- **Ticket:** [sc173901](https://app.shortcut.com/genepi/story/173901)
- **Env:** [https://location-dropdown-frontend.dev.genepi.czi.technology/upload/1](https://location-dropdown-frontend.dev.genepi.czi.technology/upload/1)

### Demos:

Screen Shot 2021-11-24 at 16 10 50
Screen Shot 2021-11-24 at 16 11 31
Screen Shot 2021-11-23 at 16 32 55
Screen Shot 2021-11-23 at 16 37 03
Screen Shot 2021-11-23 at 16 48 49

### Checklist:
- [x] I merged latest `<base branch>`
- [x] I manually verified the change
- [x] I added labels to my PR
- [x] I tested in multiple browsers
- [x] I added relevant unit tests
- [ ] I have notified others of changes they need to make locally (migrations, jobs, package updates, etc)

### CODE REVIEW

File: src/frontend/src/common/types/location.d.ts [hunk 1]
```diff
@@ -0,0 +1,15 @@
+interface GisaidLocation {
+  [index: string]: string | number;
```
Comment (mmmmmmaya) [inline src/frontend/src/common/types/location.d.ts]: why did you add this to each type?

File: src/frontend/package.json [hunk 2]
```diff
@@ -33,7 +33,7 @@
     "@typescript-eslint/parser": "^4.5.0",
     "babel-eslint": "^10.1.0",
     "classnames": "^2.2.6",
-    "czifui": "^1.2.1",
+    "czifui": "^1.4.0",
```
Comment (mmmmmmaya) [inline src/frontend/package.json:36]: did you consider pinning the latest version (1.5.0)?

File: src/frontend/src/views/Upload/components/Metadata/components/Table/components/Row/components/LocationField/index.tsx [hunk 3]
```diff
@@ -1,75 +1,126 @@
...
-
-const SELECT_PROPS = {
-  displayEmpty: true,
-  renderValue,
-};
+import { StyledDiv } from "./style";
 
 interface Props {
   fieldKey: keyof Metadata;
+  accessoryKey: keyof Metadata;
```
Comment (mmmmmmaya) [inline src/frontend/src/views/Upload/components/Metadata/components/Table/components/Row/components/LocationField/index.tsx]: what does this prop represent, conceptually?

File: src/frontend/src/common/api/index.tsx [hunk 4]
```diff
@@ -13,6 +13,7 @@ export enum API {
   GET_FASTA_URL = "/api/sequences/getfastaurl",
   USHER_TREE_OPTIONS = "/api/usher/tree_options",
   SAMPLES_VALIDATE_IDS = "/api/samples/validate-ids",
+  LOCATIONS = "/v2/locations/",
```
Comment (mmmmmmaya) [inline src/frontend/src/common/api/index.tsx:16]: I noticed this is the only endpoint with a trailing forward slash. Is that intentional?

**Comment by mmmmmmaya** [comment]
As a general note, I wish this PR had been split up into at least 2-3 smaller PR. I had some trouble keeping all of the information about what was changed in my mind given the size of the changes here.

File: .happy/terraform/modules/sfn_config/gisaid.wdl [hunk 5]
```diff
@@ -316,6 +325,39 @@ task ImportGISAID {
...
+    command <<<
+    set -Eeuo pipefail
+    aws configure set region ~{aws_region}
+
+    export ASPEN_CONFIG_SECRET_NAME=~{aspen_config_secret_name}
+    if [ "~{remote_dev_prefix}" != "" ]; then
+        export REMOTE_DEV_PREFIX="~{remote_dev_prefix}"
+    fi
+
+    export PYTHONUNBUFFERED=true
```
Comment (phoenixAja) [inline .happy/terraform/modules/sfn_config/gisaid.wdl:359]: small nit but i feel like setting this to `false` makes more sense since we want stdout and stderr streams to be unbuffered

**Comment by danrlu** [comment]
I just tried this out, this is SO NICE TnT thank you guys!!

File: src/backend/aspen/workflows/import_locations/import_locations.sh [hunk 6]
```diff
@@ -0,0 +1,25 @@
+#!/bin/bash
+
+# WDL inputs available through environmental vars:
+# AWS_REGION
+# ASPEN_CONFIG_SECRET_NAME
+# REMOTE_DEV_PREFIX (if set)
+# WORKFLOW_ID
+# S3_FILESTEM
```
Comment (phoenixAja) [inline src/backend/aspen/workflows/import_locations/import_locations.sh]: can we remove this?

File: src/backend/aspen/workflows/import_locations/save.py [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+    if test:
+        print("Success!")
+        return
+
+    interface: SqlAlchemyInterface = init_db(get_db_uri(Config()))
+    with session_scope(interface) as session:
+        # Insert all locations from gisaid_metadata
+        insert_rows_from_gisaid = "INSERT INTO aspen.locations (region, country, division, location) SELECT DISTINCT region, country, division, location FROM aspen.gisaid_metadata ON CONFLICT (region, country, division, location) DO NOTHING"
+        # Make sure we have equivalent rows for region/country/division but with null locations
+        insert_rows_with_null_locations = "INSERT INTO aspen.locations (region, country, division) SELECT DISTINCT region, country, division FROM aspen.locations EXCEPT SELECT DISTINCT region, country, division FROM aspen.locations WHERE location IS NULL"
```
Comment (phoenixAja) [inline src/backend/aspen/workflows/import_locations/save.py:26]: why did you decide to use raw sql here instead of sqlalchemy api? just curious

File: src/backend/database_migrations/versions/20211117_213611_add_locations_table.py [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+    # X    just the list of existing sample locations from gisaid
+    # 2a. Update the gisaid import job to keep this list up to date in the db.
+    # 4. Update Orange County and Rickettsial lab samples to point to a real location
+    # 5. Fix the test samples for the admin group
+    # 6. Update all the existing sample rows in the DB to match a real location
+    # 7. Create an API endpoint to return a list of locations
+    # 8. Update upload form to be searchable textbox for location api endpoint
+    # 9. Update upload form (or upload backend endpoint????) to find the appropriate location id to associate with a sample
+    # --- AFTER the above is deployed ---
+    # 10. remove sample, division, location fields from samples table, make sure the backend is 100% using associated location info.
```
Comment (phoenixAja) [inline src/backend/database_migrations/versions/20211117_213611_add_locations_table.py]: i think we can also remove this? I do enjoy seeing your process though lol

File: src/frontend/src/common/types/location.d.ts (hunk 1 above)
Comment (lvreynoso) [reply src/frontend/src/common/types/location.d.ts]: I did this because I was accessing object properties via string variables, e.g. [this function here](https://github.com/chanzuckerberg/aspen/blob/f144494f4e9b88f46052c3a15fa3f1125160c464/src/frontend/src/views/Upload/components/Metadata/index.tsx#L44-L59)

File: src/frontend/package.json (hunk 2 above)
Comment (lvreynoso) [reply src/frontend/package.json:36]: I am happy to pin the latest version, `1.4.0` was just the latest when I changed it.

File: src/frontend/src/views/Upload/components/Metadata/components/Table/components/Row/components/LocationField/index.tsx (hunk 3 above)
Comment (lvreynoso) [reply src/frontend/src/views/Upload/components/Metadata/components/Table/components/Row/components/LocationField/index.tsx]: That...I don't think we need any more. I'll fix it.

File: src/frontend/src/common/api/index.tsx (hunk 4 above)
Comment (lvreynoso) [reply src/frontend/src/common/api/index.tsx:16]: Yes, our FastAPI setup is particular about having a trailing slash in the url.

File: .happy/terraform/modules/sfn_config/gisaid.wdl (hunk 5 above)
Comment (lvreynoso) [reply .happy/terraform/modules/sfn_config/gisaid.wdl:359]: If we want the streams to be unbuffered, then we should keep this `true`. (The var is PYTHON + **UN**BUFFERED).

File: src/backend/aspen/workflows/import_locations/save.py (hunk 7 above)
Comment (lvreynoso) [reply src/backend/aspen/workflows/import_locations/save.py:26]: @jgadling do you want to take this? I know when we first reasoned about how to do this we were thinking in SQL syntax

**Comment by lvreynoso** [comment]
@mmmmmmaya I agree, we really should have split up this PR. Apologies for that, I'll make sure in the future PRs are more limited in scope.

File: .happy/terraform/modules/sfn_config/gisaid.wdl (hunk 5 above)
Comment (phoenixAja) [reply .happy/terraform/modules/sfn_config/gisaid.wdl:359]: oh whoops i read that as `PYTHONBUFFERED`, my mistake

File: src/backend/aspen/database/models/locations.py [hunk 9]
```diff
@@ -0,0 +1,15 @@
+from sqlalchemy import Column, String
+
+from aspen.database.models.base import idbase
+from aspen.database.models.mixins import DictMixin
+
+
+class Location(idbase, DictMixin):  # type: ignore
```
Comment (phoenixAja) [inline src/backend/aspen/database/models/locations.py:7]: do we need `DictMixin` if we're using the serializers now?

File: src/backend/database_migrations/versions/20211117_213611_add_locations_table.py [hunk 10]
```diff
@@ -0,0 +1,58 @@
+"""Add Locations Table
+
+Create Date: 2021-11-17 21:36:11.102820
+
+"""
+import enumtables  # noqa: F401
+import sqlalchemy as sa
+from alembic import op
+
+# revision identifiers, used by Alembic.
```
Comment (phoenixAja) [inline src/backend/database_migrations/versions/20211117_213611_add_locations_table.py:10]:
```suggestion
```

**Review by phoenixAja — APPROVED** [review]
🤩 🤩 🤩 so awesome!! Great work @lvreynoso

File: src/backend/aspen/workflows/import_locations/save.py (hunk 7 above)
Comment (jgadling) [reply src/backend/aspen/workflows/import_locations/save.py:26]: Know what? I just assumed SQLAlchemy didn't have a construct for `insert ... select` or `on conflict` but it turns out that it does:
https://docs.sqlalchemy.org/en/14/core/dml.html?highlight=from_select#sqlalchemy.sql.expression.Insert.from_select
https://docs.sqlalchemy.org/en/14/dialects/sqlite.html?highlight=on_conflict#sqlalchemy.dialects.sqlite.Insert.on_conflict_do_nothing

I even see an except clause:
https://docs.sqlalchemy.org/en/14/core/selectable.html?highlight=except#sqlalchemy.sql.expression.except_

I'll see about translating these queries to SA code in the morning.

File: src/backend/aspen/app/views/sample.py [hunk 11]
```diff
@@ -447,6 +445,23 @@ def create_sample():
...
+                g.db_session.query(Location)
+                .filter(Location.id == data["sample"].get("location_id", None))
+                .one_or_none()
+            )
+            if valid_location:
+                sample_args["location_id"] = valid_location.id
+                sample_args["region"] = RegionType(valid_location.region)
+                sample_args["country"] = valid_location.country
+                sample_args["division"] = valid_location.division
+                sample_args["location"] = valid_location.location or ""
```
Comment (vincent-czi) [inline src/backend/aspen/app/views/sample.py:458]: There's currently a bug -- well, it could be intended, but I don't think it is -- regarding what happens when a user uploads a non-location-specific "location" for a sample's metadata. This is a screenshot of a Samples entry in table from the rdev when the user selects `North America/USA/California` as the Collection Location during its metadata upload:
image
The `---` is where the Current Location should be. I believe this line is the specific cause, but not the root issue. The cause is because, when submitting `North America/USA/California`, that has no `location` aspect (it's `division` level only), so when the Samples table pulls that sample, there is an empty string for the `location`, so that gets parsed as false-y, which triggers the table cell renderer in FE to show it as `---`. Previously, submitting "California" as the collection location actually set the samples table `location` column to `California`. But now since we're truly matching the correct GISAID location, the lowest level `location` has no value for a general division. We need some kind of fallback for cases like this.

More generally though, I think there's a root issue of conflicting sources of truth regarding location. While the `samples` table still has `region`, `country`, `division`, and `location`, the addition of now having a FK `location_id` means we can have conflict between the sample's location info and the canonical location info provided by looking it up in the `locations` table. I don't have an easy answer for how to fix this -- the migration for legacy data could be hairy -- but I think we probably don't want two ways to look up the location.

I think we should either:
1. Not have a `location_id` for the samples. While the available options for location data (when user is submitting the metadata via dropdown, TSV, etc) do come from the `samples` table, they are just used as a way to populate the `samples` table. **Pro:** Easily backwards compatible with our legacy approach; **Con:** If the GISAID canonical names for locations drift, this could cause us pain. Also adds a bit more data storage overhead since we're duplicating values across many samples that could be consolidated.
2. Not have any of `region`, `country`, `division`, and `location` in the samples table, only have a `location_id` that is a FK to pulling the canonical values from that table. **Pro:** Normalized data, keeps us using canonical terms from GISAID; **Con:** Lots more code updates required, probably a reasonably painful DB migration to move old sample locations to new format, we need to watch GISAID locations for drift -- I've only skimmed the PR, don't really know how we ingest the GISAID locations, but if we're using their IDs for locations, we have to make sure that never changes under our feet or it could screw up all our locations so far.

I think (2) is the more "right" way to do it, but it's definitely more of a hassle. Also ties us more heavily to GISAID than we might want long-term if we're ever needing to provide a different set of locations for a non-COVID pathogen in the future. But as it stands, I'm not really sure what should happen if there's ever a conflict between the location info in a sample row and the `locations` row referenced by `location_id`.

File: src/frontend/src/views/Upload/components/Metadata/components/Table/components/Row/components/LocationField/index.tsx [hunk 12]
```diff
@@ -1,75 +1,126 @@
...
+    dispatch({ searching: true });
+
+    const regex = new RegExp(escapeRegExp(query), "i");
+    const filteredLocationOptions = locationOptions.filter((option) =>
+      regex.test(option.name)
+    );
+    // alphabetical sort
+    // this ensures partial locations (i.e. region, country and divison
+    // but no location) end up on top.
+    const sortedLocationOptions = filteredLocationOptions.sort((a, b) => {
```
Comment (mmmmmmaya) [inline src/frontend/src/views/Upload/components/Metadata/components/Table/components/Row/components/LocationField/index.tsx:68]: I think this is the default sort behavior (ie, you can just say filteredLocationOptions.sort() instead of defining the whole thing here).

**Comment by Janeece** [comment]
Wow this is awesome! 

Here's design feedback:

- [x] **Should Have** - Can we use the dropdown use the InputDropdown style dropdown instead of the minimal dropdown?
Screen Shot 2021-11-29 at 2 49 36 PM

- [x] **Should Have** - When reopening the menu from a selected location, can we show the selected location in the menu? 
Screen Shot 2021-11-29 at 3 25 00 PM

- [ ] **Nice to Have** - It feels like we need a max width/design for content overflow for this input. I can think about that later in the week but this is not a blocker. 


- Given the feedback above, i thought a design prototype of how we want this to look would be helpful: https://www.figma.com/proto/pH5rRGCzcMkCKkn6UMQVUR/Upload?node-id=1265%3A62978&scaling=min-zoom&page-id=1265%3A61639
- **Question:** how does this info show up on the sample table? Just county or the full location name?

File: src/backend/aspen/app/views/sample.py (hunk 11 above)
Comment (jgadling) [reply src/backend/aspen/app/views/sample.py:458]: @vincent-czi just chiming in here -- #2 is where we're going, but we still have a fair bit of code that depends on the country/division/location/etc fields in the samples table, and this PR is already massive.

Our YAML builder project will require us to touch most of the code that works with sample locations, so the plan is to deprecate these old fields as part of that project.

File: src/frontend/src/common/types/location.d.ts (hunk 1 above)
Comment (lvreynoso) [reply src/frontend/src/common/types/location.d.ts]: Ended up removing this.

File: src/frontend/src/views/Upload/components/Metadata/components/Table/components/Row/index.tsx [hunk 13]
```diff
@@ -27,7 +30,16 @@ const validationSchema = yup.object({
     .min(10, DATE_ERROR_MESSAGE)
     .max(10, DATE_ERROR_MESSAGE)
     .required("Required"),
-  collectionLocation: yup.string().required("Required"),
+  collectionLocation: yup
+    .object({
+      id: yup.number().required(),
+      // region: yup.string(),
```
Comment (mmmmmmaya) [inline src/frontend/src/views/Upload/components/Metadata/components/Table/components/Row/index.tsx:36]: why are these commented out? are they required?

File: src/backend/aspen/app/views/sample.py (hunk 11 above)
Comment (vincent-czi) [reply src/backend/aspen/app/views/sample.py:458]: Okay, as long as there's a clear direction and it will be happening in the next couple months I think that's fine.

I do think we probably want some kind of helper func that can take either a Locations record or a dict of region, country, division, and location, then just return the first existing field of the most specific kind. That way we can solve for the "California" bug above.

File: src/frontend/src/views/Upload/components/Metadata/index.tsx [hunk 14]
```diff
@@ -30,14 +33,67 @@ import Table from "./components/Table";
...
+  });
+  const candidateLocation = scoredLocations.reduce(
+    ([prevLocation, prevScore], [currLocation, currScore]) => {
+      if (currScore < prevScore) {
+        return [currLocation, currScore];
+      }
+      return [prevLocation, prevScore];
+    }
+  );
+  console.log(`Found match for ${locationString}:`);
```
Comment (mmmmmmaya) [inline src/frontend/src/views/Upload/components/Metadata/index.tsx:92]: these logs should come out before we commit

File: src/frontend/src/views/Upload/components/Metadata/components/Table/components/Row/components/LocationField/index.tsx [hunk 15]
```diff
@@ -1,75 +1,119 @@
...
+  };
+
+  return (
+    <StyledDiv onBlur={handleBlur}>
+      <Dropdown
+        label={value?.name || "Search For Location"}
+        value={value}
+        onChange={handleLocationChange}
+        options={state.results}
+        search={true}
```
Comment (mmmmmmaya) [inline src/frontend/src/views/Upload/components/Metadata/components/Table/components/Row/components/LocationField/index.tsx:104]:
```suggestion
        search
```

in React, `propName={true}` is equivalent to `propName` with no value specified

File: src/frontend/src/views/Upload/components/Metadata/components/Table/components/Row/components/LocationField/index.tsx [hunk 16]
```diff
@@ -1,75 +1,119 @@
...
-            value={value}
-            handleClick={applyToAllColumn}
-          />
-        ))
+  const [state, dispatch] = useReducer(locationSearchReducer, {
+    results: [],
+    searching: false,
+  });
+
+  const searcher = async (query: string): Promise<void> => {
```
Comment (mmmmmmaya) [inline src/frontend/src/views/Upload/components/Metadata/components/Table/components/Row/components/LocationField/index.tsx:54]: why did you decide to manage the search state here instead of relying on the built-in search behavior in the Dropdown?

**Comment by lvreynoso** [comment]
Closing to split this into 2 PRs. Will continue to work on feedback.
